Add systemstart task scripts to installer repo, fix bug

MCGallaspy opened the pull request:
Fixes #106 -- should start at boot now.
These scripts are updated, and should be removed from the ka-lite repo... they do no good there!

commitmsg
```
Copy them, and instead of referencing the now defunct bin/windows/kalite.bat
script, use the included kalite.bat script in the ..\Python27\Scripts\ dir.
```

MCGallaspy commented:
Check out [this build](http://dungeon.learningequality.org:8085/browse/KL-KALITE016-WIN-87) to test it out!

MCGallaspy commented:
@radinamatic let's give it a shot

radinamatic commented:
Gottit! Downloading...

radinamatic commented:
No luck, @MCGallaspy, at least on Windows 8 (testing 7 now):

- I still get "Failed to create config file/option" alerts
- Scheduled task is created correctly, and at least according to system, it is executed correctly too, but KA Lite does not run at startup.
 (screenshot coming)

radinamatic commented:
Windows XP:

- No "Failed to create config file/option" alerts
- Scheduled task is created correctly, but KA Lite does not run at startup. 

![wxp](https://cloud.githubusercontent.com/assets/1457929/11898369/9fcf468a-a597-11e5-8c3d-cebc8b213b1e.png)

radinamatic commented:
However, if I activate all three options and restart the machine, both KA Lite and server are started! :)

MCGallaspy commented:
> I still get "Failed to create config file/option" alerts

That's expected -- this only addresses the "start at boot" option

radinamatic commented:
Windows 7 yields the same as result as 8:  Scheduled task is created correctly, but KA Lite does not run at startup.

(screenshots of some task properties coming)

radinamatic commented:
However, as noted before for XP, if I activate the "Run KA Lite when the user logs in" and restart, that task gives us the desired result (KA Lite is run at startup). Could this be a user problem?

MCGallaspy commented:
> if I activate the "Run KA Lite when the user logs in" and restart, that task gives us the desired result (KA Lite is run at startup). 

Does it truly run when the system starts? Or just when that user logs in? I'm specifically trying to debug the "at system start" option. If the "when user logs in" options works ok, try logging in as a different user and seeing if it starts -- I do not expect it to.

radinamatic commented:
You are probably right, I'll create another user (this one came with VM image by default), and check that out.

MCGallaspy commented:
I'm attempting to install & replicate, but in the meantime it could be informative to run the script which the scheduled task is configured to run, to see if there is any error output.

radinamatic commented:
Running (double clicking) `kalite.bat` opens and immediately closes the command prompt window, so to be able to see the output, I made a copy of the `.bat` file, edited it in Notepad adding `@pause` at the end and run it again. This is a output:

![selection_120](https://cloud.githubusercontent.com/assets/1457929/11906867/f28fdf28-a5cf-11e5-8576-fede7825efd1.png)

MCGallaspy commented:
yeah, `kalite.bat` is meant to be run from the command line

MCGallaspy commented:
Believe I know why, @radinamatic if you can confirm. During installation, we create the scheduled task using the script from this PR, which relies on expanding the variable %KALITE_SCRIPT_DIR%. But that variable is not expanded because the environment will not be updated until the setup exe exits. The result is a misconfigured scheduled task:

![virtualbox_ie9 - win7_18_12_2015_14_04_20](https://cloud.githubusercontent.com/assets/8888020/11908381/93176596-a590-11e5-85f3-7fdfe7ecfc4a.png)

Note that the full path to the script is missing. Radina, can you confirm that post-installation, having checked the "start at system startup" option, that the scheduled task is misconfigured the same way it is here?

radinamatic commented:
My above "Actions" screenshot from W7 seems identical @MCGallaspy.
I restarted the W8 VM and this is what's inside the KA Lite task:

![ie11 - win8 1 py 3 5 1 first running - oracle vm virtualbox_121](https://cloud.githubusercontent.com/assets/1457929/11908804/abf905e8-a5df-11e5-9262-c474021e587c.png)

I'm gonna revert and do another fresh install, just in case I meddled with the above one.

radinamatic commented:
After fresh install on W8, it seems the same as in your example @MCGallaspy:

![ie11 - win8 1 pre ka lite running - oracle vm virtualbox_124](https://cloud.githubusercontent.com/assets/1457929/11909210/d7b07ffa-a5e3-11e5-9707-87ec2b3a8497.png)

MCGallaspy commented:
Yeah, I see. I just submitted a patch, should be the latest build on bamboo. Let's try installing that one "fresh".

radinamatic commented:
Ok, tomorrow I'll be testing build # 90 on all the machines, and report back! @MCGallaspy

MCGallaspy commented:
TODO: once this is merged, I need to modify the repos used on the build server.

MCGallaspy commented:
Looking more closely at your screenshots, there's some bizarre options for the tasks (like only running when on AC power). Seems to be the default, however... Will address those here as well.

radinamatic commented:
Win XP tested today with build `#91`, still no luck with start on boot... :(

![xp-build91](https://cloud.githubusercontent.com/assets/1457929/11937847/26075ebe-a817-11e5-9ad3-14ef0f848909.jpg)

radinamatic commented on windows/installer-source/KaliteSetupScript.iss:
```diff
@@ -444,7 +444,7 @@ begin
             if StartupPage.SelectedValueIndex = 0 then
             begin
-                if ShellExec('open','guitools.vbs','4', ExpandConstant('{app}'), SW_HIDE, ewWaitUntilTerminated, StartupCode) then
+                if Exec(ExpandConstant('{cmd}'), '/C "schtasks /create /tn "KALite" /tr "\"{reg:HKLM\System\CurrentControlSet\Control\Session Manager\Environment,KALITE_SCRIPT_DIR}\kalite.bat\" start" /sc onstart /ru SYSTEM"', '', SW_HIDE, ewWaitUntilTerminated, StartupCode) then
```
I was trying to understand this line of code and run into this, maybe it could be of help...?

http://stackoverflow.com/questions/3636055/how-to-modify-the-path-variable-definitely-through-the-command-line-in-windows

MCGallaspy replied:
The values in the curly braces should have been replaced by IS with the given registry value -- but it seems it wasn't.

radinamatic commented:
Ok, I've been at this for a while now, trying various options why the task may have `(0x1)` as "Last Run Result" as listed **[here](https://social.technet.microsoft.com/Forums/windowsserver/en-US/57c65328-d8c8-4e13-9b51-e017ae6c298c/task-scheduler-returns-0x1-code-when-run-whehter-user-is-logged-on-or-not-options-selected?forum=winserverManagement)**, when I realized I needed to clarify something, mainly to myself.

If the user selects `Run the server at system startup`, that would imply that, if the scheduled task is working correctly, `python` process is activated on startup, but not `KA Lite.exe`. This has been failing on all the systems so far. 

I know it doesn't help much since you are trying to work out the solution with the `KALITE_SCRIPT_DIR`, but I did manage to start the `python` process by putting the local path `C:\Python27\Scripts` in both "Program/Script" and "Start in (optional)" options of the "Actions" tab, the second one without the quotes:

Bad news is that even with this `python` process running and listening to the correct port, I still cannot open KA Lite in the browser: it reports "Connecting..." and "Waiting for 127.0..0.1..."

(screenshot coming)

Could it have something to do with the "Last Run Result" now being `The task is currently running (0x41301)` and **not** being `successful`?

Another issue is that once user starts `KA Lite.exe` itself from the desktop shortcut (if they choose to add it), the second option `Options > Run KA Lite at system startup` seems be activated by default, but it is not working, namely, `KA Lite.exe` does not  run on startup:

![boot3](https://cloud.githubusercontent.com/assets/1457929/11945401/d54a4f48-a84f-11e5-9b64-eb1dd18508d0.png)

This possibly deserves a separate issue.

radinamatic commented:
Correction, server did start at the end, but it took good 10-15 minutes to do so... :P
I'll retest this tomorrow on a another VM with more RAM, maybe that was the issue. 

So, maybe the next step is to figure out how to pull that directory from the registry and put it in the "Start in (optional)" option too? I tried on my end, but given that it's a registry path (with blank spaces in between on top of everything), I couldn't figure out the right approach. 
@MCGallaspy

radinamatic commented:
The above test case (putting the local path `C:\Python27\Scripts` in both "Program/Script" and "Start in (optional)" options of the "Actions" tab, the second one without the quotes), worked in Windows 8.1 too.

(screenshot coming)

I wasn't able to test it on Windows XP because the Task Scheduler is a bit more "primitive", and is picky in allowing changes.

radinamatic commented:
Regarding XP, though, seems that it's the only one that fills in the "Start in:" option of the "Actions" tab by default, from the script that you have in the installer, while Windows 7 & 8 have only the "Program/Script" option after the installation. So, a part from figuring out why that reg:HKLM fails to grab the value of  `KALITE_SCRIPT_DIR` env var (and I did check, **it is in the registry**), try forcing the "Start in:" option, because it is mentioned on several places that I checked as a potential cause of scheduled task fail.

MCGallaspy commented:
Looks like you've been busy, so I'm addressing comment by comment:

Re: https://github.com/learningequality/installers/pull/322#issuecomment-166383187

The values enclosed in `{` & `}` should be expanded by IS -- but it seems they weren't. I'll double check. After installation finishes, what is the value of that registry key? You can check using `regedit`, and the path is given, i.e. the root key is HKLM, then System, etc, and you can follow it until you get to Environment.

radinamatic commented:
If I recall well, the value of the registry key was the (correct) Python scripts path on that machine.

I'll fire up the VM and recheck.

MCGallaspy commented:
I read up, looks like the registry value is set properly -- is it substituted correctly in the scheduled task on other version of Windows? In other words, does it fail to substitute only on XP?

You also mentioned that you added C:\Python27\Scripts to your PATH, and the scheduled task ran properly -- did you edit the scheduled task prior to changing the PATH value? In other words, there are four cases. Can you fill in the values in the table below?


|      | bad schtask | good schtask          |
| ---- | --------- | ----------- |
|    **Scripts dir in PATH**   |   ?      | ? .|
|   **Scripts dir *not* in PATH**   | Does not work     |  ?     |

radinamatic commented:
**Yes**, registry value is set correctly on all versions (XP, 7, 8), and **No**, it's not substituted correctly in scheduled task, so **Yes**, task fails on all the versions.

I did not edit/add Python to the path, installer does that correctly, I just edited the task itself, substituting the unexpanded `{...}` in the task with `C:\Python27\Scripts`.

(screenshot coming) 

It might be in one of the links that I posted above, but there might have been a suggestion somewhere that in order for it to expand correctly, you must set the reg value type as `REG_EXPAND_SZ` instead of `REG_SZ`.

radinamatic commented:
Addendum to the above comment: `C:\Python27\Scripts` is added to the PATH **IF** the user ticks the appropriate checkbox during the installation. If they do not, or Python 2 is the sole one installed, it's not the problem, as it is taken by default. This only becomes a problem when there is a Python 3 in the mix too, and its path **is** added (see #296).

MCGallaspy commented:
Okay, this latest commit should expand the constant if you picked the "Start at boot" option initially. We can test it out from this build: http://dungeon.learningequality.org:8085/browse/KL-KALITE016-WIN-93

radinamatic commented:
Good news @MCGallaspy, so far it seems to be working on Windows 7, raising the Python process correctly... :)

I'll test on rest of the versions and report back!

radinamatic commented:
We are not out of the woods just yet... :(

Build # 93 Worked flawlessly on Windows 7 & 8, but not on XP. Scheduled task seems OK (`C:\Python27\Scripts` is expanded correctly), but Python process is not raised, and server never starts. `kalite.bat` works when run from Terminal. 

I've been at this for a while, can't figure out why. If you have any ideas, yell!

MCGallaspy commented:
It may be the user that the scheduled task is run as (SYSTEM) -- if it works when you run as yourself, can you try opening a cmd.exe prompt as SYSTEM (this option should be available from the "Run as..." context menu when you right click) then running kalite.bat to see what happens?

MCGallaspy commented:
Yeah, after experimenting with it I think the solution is to run the scheduled task as a different user on XP only -- looking in to it.

radinamatic commented:
If I try to open cmd.exe prompt from the "Run as...", I don't get to choose SYSTEM, it's not available as an option. There is only "Administrator" user alongside the default VM user, and if I choose it, it demands a password that I'm unaware of (VM only offers password for the default user).

Scheduled tasks log shows the following:
```
"KALite.job" (kalite.bat)
	Finished 12/30/2015 8:29:37 AM
	Result: The task completed with an exit code of (1).
```

`1 or 0x1` exit code implies incorrect or unknown function called, if I recall well. Could it be the that the argument `start` is not passed correctly? In XP it does not have a separate field in the task properties window as for 7 or 8.

MCGallaspy commented:
I'm not sure what the error code precisely means without looking at the output of the script, which I'm not really sure how to do. However, I found that when the task is run as a different user (not SYSTEM) then it works without error.

radinamatic commented:
True, if I change the scheduled task (or make a new one) with `Run as:` set for the default VM user, task is executed correctly, python process is raised and server starts on boot no prob...

So does that mean that for XP we'll have to configure the installer to run `kalite.bat` task set for the logged-in user?

MCGallaspy commented:
@radinamatic okay, this should start the task in Windows XP. For XP only, the scheduled task is set to run as the installing user -- in this case it should be the admin, since the installer is run with elevated privileges. I'm afraid this might make upgrading on XP a little tricky, since it may mean *moving* files from one location to another. But for later versions of Windows, this should introduce no change. If this works well for a regular install, then I'll merge this and we can begin testing upgrade scenarios, with an emphasis on XP.

radinamatic commented:
Ok, waiting on build # 94 to test this @MCGallaspy.

radinamatic commented:
Build #94 seems to be problematic only on XP, on Windows 7 installer finishes correctly, and starts the server Ok on restart.

MCGallaspy commented:
Yeah, on XP it prompts the user for a password.

MCGallaspy commented:
I'm disabling the start at boot option from the task tray program on XP. Shortly I'll remove the option from the installer.

MCGallaspy commented:
@radinamatic I'm merging this so we can start working on #321.
